File: solutions/fsharp/queen-attack/3/QueenAttack.fs
module QueenAttack

let create (position: int * int) =
   let onBoard n = 0 <= n && n <= 7
   let column, row = position
   
   onBoard column && onBoard row

let distance (lhs:int) (rhs:int) = abs (lhs - rhs)

let canAttack (queen1: int * int) (queen2: int * int) =
   let row1, column1 = queen1
   let row2, column2 = queen2
   
   let attacksOnRow =
      row1 = row2
   let attacksOnColumn =
      column1 = column2
   let attacksOnDiagonal =
      let distanceX = distance row1 row2
      let distanceY = distance column1 column2
      distanceX = distanceY

   attacksOnRow || attacksOnColumn || attacksOnDiagonal

File: solutions/csharp/queen-attack/3/QueenAttack.cs
public class Queen
{
    const int MinPosition = 0;
    const int MaxPosition = 7;

    public Queen(int row, int column)
    {
        if (row < MinPosition || row > MaxPosition ||
            column < MinPosition || column > MaxPosition)
            throw new ArgumentOutOfRangeException();

        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }

    public bool SharesRowWith(Queen other) => Row == other.Row;
    public bool SharesColumnWith(Queen other) => Column == other.Column;
    public bool SharesDiagonalWith(Queen other) =>
        Math.Abs(Row - other.Row) == Math.Abs(Column - other.Column);
}

public static class QueenAttack
{
    public static Queen Create(int row, int column) => new(row, column);

    public static bool CanAttack(Queen white, Queen black) =>
        white.SharesRowWith(black) ||
        white.SharesColumnWith(black) ||
        white.SharesDiagonalWith(black);
}